### PR TITLE
CA-223473: Change pool_apply to handle host apply failure

### PR DESCRIPTION
**Pull request by cheng-z**
Signed-off-by: Cheng Zhang cheng.zhang@citrix.com

### CODE REVIEW

File: ocaml/xapi/xapi_pool_update.ml [hunk 1]
```diff
@@ -311,12 +311,16 @@ let introduce ~__context ~vdi =
...
-      (fun rpc session_id -> Client.Pool_update.apply rpc session_id self host))
-  )
+  List.fold_left (fun acc host ->
+    try
+      ignore(Helpers.call_api_functions ~__context
+        (fun rpc session_id -> Client.Pool_update.apply rpc session_id self host));
+      acc
+    with e -> (Db.Host.get_name_label ~__context ~self:host)::acc
+  ) [] in
+  if List.length failed_hosts > 0 then raise (Api_errors.Server_error(Api_errors.update_pool_apply_failed, failed_hosts))
```
Comment (jonludlam) [inline ocaml/xapi/xapi_pool_update.ml]: It's more common in exceptions to return references rather than the name_label.

**Comment by jonludlam** [comment]
Looks OK, except please change the names -> references in the exception.

File: ocaml/xapi/xapi_pool_update.ml (hunk 1 above)
Comment (cheng-z) [reply ocaml/xapi/xapi_pool_update.ml]: Thanks @jonludlam 
Code is now modified, please help to review again.
Thanks.

File: ocaml/xapi/xapi_pool_update.ml [hunk 2]
```diff
@@ -311,12 +311,16 @@ let introduce ~__context ~vdi =
...
-  List.iter (fun host ->
-    ignore(Helpers.call_api_functions ~__context
-      (fun rpc session_id -> Client.Pool_update.apply rpc session_id self host))
-  )
+  List.fold_left (fun acc host ->
+    try
+      ignore(Helpers.call_api_functions ~__context
+        (fun rpc session_id -> Client.Pool_update.apply rpc session_id self host));
+      acc
+    with e -> host :: acc
```
Comment (jonludlam) [inline ocaml/xapi/xapi_pool_update.ml]: One more thing: we should log the error here so it turns up in xensource.log.

**Comment by cheng-z** [comment]
Hi @jonludlam 
Code reworked.
Could you please help to review again?
Thanks.

**Pull request merged**